### ISSUE TITLE
503 when timeout in the queue ends with empty reply from server

### ISSUE THREAD

**Issue by vvucetic** [haproxy/haproxy#2404]
### Detailed Description of the Problem

When server queue is full, request ends with empty reply from HAProxy, logged as 503 and status `sQ`. That happens in `2.6.16-alpine` as well as all never versions.

### Expected Behavior

In the version `2.6.15-alpine` however, request is properly accepted in the queue regardless if the server queue is full and stays in the, I believe, backend queue until `timeout queue` and then responds with 503 error page (but not empty response)

Something changed between the versions and it's easy to reproduce.

### Steps to Reproduce the Behavior

1. Make backend with 1 server with maxconn 1.
2. Make sure that the server responds slowly so it gives you time to send another request
3. Make one request which will take the connection spot. Make another immediately after. 

In version `2.6.15-alpine` you'll get 503 after set timeout. In version `2.6.16-alpine` the second request will immediately end up with an empty response. I can share docker-compose with mock service if needed. 

### What is your configuration?

```haproxy
global
    daemon
    maxconn 10000
    log stdout format raw local0 debug
    tune.lua.session-timeout 40ms

defaults
    mode http
    log global
    option httplog
    timeout connect 5s
    timeout client 91s
    timeout server 91s
    maxconn 8000
    option forwardfor
    option abortonclose
    retry-on conn-failure 504
    compression algo gzip
    compression type text/html text/plain application/json

frontend http
    mode http
    bind :8080
    tcp-request inspect-delay 5s

    http-request set-var(txn.rl_backend_company) str(unknown)
    http-after-response set-header x-haproxy-rl-backend-company %[var(txn.rl_backend_company)]

    use_backend default

backend default
    maxconn 10000
    timeout queue 10ms
    mode http

    default-server init-addr last,libc,none check inter 1s resolvers docker maxconn 1
    server s1 service:6000

resolvers docker
    nameserver docker 127.0.0.11:53
    accepted_payload_size 8192 # allow larger DNS payloads
```

### Output of `haproxy -vv`

```plain
(shared exact versions)
```

**Comment by vvucetic** [comment]
If you want to easily reproduce, here are additional mock files to set up environment

## Mock service
`service_mock/dockerfile`
```
FROM python:3.7-alpine

RUN mkdir /app

WORKDIR /app

ADD requirements.txt /app

ADD main.py /app

RUN pip3 install -r requirements.txt

CMD ["gunicorn", "-w 4", "-b", "0.0.0.0:6000", "main:app", "--error-logfile gunicorn.error.log", "--access-logfile gunicorn.log", "--capture-output"]

```

`service_mock/main.py`
```
import time

import flask

app = flask.Flask(__name__)


@app.route('/mock/')
def hello_world():
    resp = flask.Response("Service result.")
    return resp


@app.route('/mock/slow')
def slow_response():
    time.sleep(10)
    resp = flask.Response("Slow service result.")
    return resp
```
`service_mock/requirements.txt`
```
Flask>=2.2.2
gunicorn>=19,<20
```
`docker-compose.yml`
```
version: '3.7'
services:
  haproxy:
    image: haproxy:2.6.15-alpine # >=2.6.16-alpine doesnt work
    sysctls:
      net.ipv4.ip_unprivileged_port_start: 0
    ports:
      - "8080:8080"
    expose:
      - 8080
    volumes:
       - ./haproxy_mock/haproxy.cfg:/usr/local/etc/haproxy/haproxy.cfg

  service:
    build:
      context: ./service_mock
      dockerfile: Dockerfile
    expose:
      - 6000
    volumes:
      - ./service_mock/:/app
```

I already shared mock haproxy config. If you just change the docker tag of the haproxy, you can observe different behavior on the second request that you can make like this

`curl -v localhost:8080/mock/slow`

![image](https://github.com/haproxy/haproxy/assets/1523657/81ff295e-754b-41b4-a8c5-479187b333ec)
![image](https://github.com/haproxy/haproxy/assets/1523657/7824625b-81b0-4b70-9c7a-0e26a55eea49)

**Comment by Darlelet** [comment]
Thanks for the detailed report and for the reproducer, I was able to set up a test environment out of it.

After bisect it turns out b9c87f808288c833d14eefacb4894bdae32fbdc8 is responsible for the change of behavior.
@capflam any ideas?

**Comment by capflam** [comment]
Indeed. Thanks @Darlelet for the analysis. It seems I failed my patch. In `sc_cond_forward_shut()`, before handling the write timeout on the other side, we must take care that the shutdown is forwardable. The following patch seems to fix the issue:
```diff
diff --git a/src/stconn.c b/src/stconn.c
index b3aed9521f..74efca7efd 100644
--- a/src/stconn.c
+++ b/src/stconn.c
@@ -522,15 +522,11 @@ struct appctx *sc_applet_create(struct stconn *sc, struct applet *app)
  */
 static inline int sc_cond_forward_shut(struct stconn *sc)
 {
-       /* Forward the shutdown if an write error occurred on the input channel */
-       if (sc_ic(sc)->flags & CF_WRITE_TIMEOUT)
-               return 1;
-
        /* The close must not be forwarded */
        if (!(sc->flags & (SC_FL_EOS|SC_FL_ABRT_DONE)) || !(sc->flags & SC_FL_NOHALF))
                return 0;
 
-       if (co_data(sc_ic(sc))) {
+       if (co_data(sc_ic(sc)) && !(sc_ic(sc)->flags & CF_WRITE_TIMEOUT)) {
                /* the shutdown cannot be forwarded now because
                 * we should flush outgoing data first. But instruct the output
                 * channel it should be done ASAP.
```

**Comment by capflam** [comment]
I pushed the fix, it should be good now. Thanks !

**Comment by vvucetic** [comment]
Hey @capflam, I see some evidence that the similar thing might be happening with `http-request deny deny_status`. 

We have a case where we conditionally set 429 status and deny request if client is cached for that status. We have Varnish in front if it and during some spikes, we see Varnish reporting 503 from HAProxy while HAProxy logged 429. That kind of behavior I noticed when I was debugging this very issue. Do you see any possibility that HAProxy didn't process the request fully and return a response with headers but just closed the connection on `http-request deny`?

I tried to recreate that case locally and I can't so it must be some kind of race condition. HAProxy we're using is from docker image3 `haproxy:2.9.4-alpine`. I can open another issue, but I don't have enough data and it's quite similar in behavior with this issue.

**Comment by capflam** [comment]
Indeed, till recently, the H1 multiplexer was unable to drain the request when the server replied and shut the connection down before the end of the request. And the same was true for response returned by HAProxy itself. This was inherited from old versions. In 3.0, the H1 request is now drained. So it is no longer an issue. But on previous version, unfortunately, the request is not drained.

**Comment by vvucetic** [comment]
> Indeed, till recently, the H1 multiplexer was unable to drain the request when the server replied and shut the connection down before the end of the request. And the same was true for response returned by HAProxy itself. This was inherited from old versions. In 3.0, the H1 request is now drained. So it is no longer an issue. But on previous version, unfortunately, the request is not drained.

Thank you!
Is there an alternative way to reject request and properly drain, until 3.0 is stable? 
Would it help if instead of `deny` we use `tarpit` so that the connection is properly drained?

**Comment by vvucetic** [comment]
On this unrelated note I asked, this is actually due to Varnish not passing on a response if a request hasn't been streamed over completely. There's open issue for years. [Linking](https://github.com/varnishcache/varnish-cache/pull/3814) here in case someone else finds it.

**Comment by capflam** [comment]
Ok, copy that. About draining of H1 requests, it will be backported to 2.9 because it is not a huge work. It will not be included to the 2.9.7, but to 2.9.8. It is in fact already available to the [2.9/next](http://git.haproxy.org/?p=haproxy-2.9.git;a=shortlog;h=refs/heads/next) branch. You may give it a try if you want.

**Comment by vvucetic** [comment]
> Ok, copy that. About draining of H1 requests, it will be backported to 2.9 because it is not a huge work. It will not be included to the 2.9.7, but to 2.9.8. It is in fact already available to the [2.9/next](http://git.haproxy.org/?p=haproxy-2.9.git;a=shortlog;h=refs/heads/next) branch. You may give it a try if you want.

I checked with version 3.0 and it's different behavior which helps with this issue. Thank you for the hint!
